vortex-dtype/src/field.rs

```diff
@@ -110,6 +126,62 @@ impl FieldPath {
         self.0 = self.0.iter().skip(1).cloned().collect();
         Some(self)
     }
+
+    /// The dtype, within the given type, to which this field path refers.
+    ///
+    /// Note that a nullable DType may contain a non-nullable DType. This function returns the
+    /// literal nullability of the child.
+    ///
+    /// # Examples
+    ///
+    /// Extract the type of the "b" field from `struct{a: list(struct{b: u32})?}`:
+    ///
+    /// ```
+    /// use std::sync::Arc;
+    ///
+    /// use vortex_dtype::*;
+    /// use vortex_dtype::Nullability::*;
+    ///
+    /// let dtype = DType::Struct(
+    ///     Arc::new(StructFields::from_iter([(
+    ///         "a",
+    ///         DType::List(
+    ///             Arc::new(DType::Struct(
+    ///                 Arc::new(StructFields::from_iter([(
+    ///                     "b",
+    ///                     DType::Primitive(PType::U32, NonNullable),
+    ///                 )])),
+    ///                 NonNullable,
+    ///             )),
+    ///             Nullable,
+    ///         ),
+    ///     )])),
+    ///     NonNullable,
+    /// );
+    ///
+    /// let path = FieldPath::from(vec![Field::from("a"), Field::ElementType, Field::from("b")]);
+    /// let resolved = path.resolve(dtype).unwrap();
+    /// assert_eq!(resolved, DType::Primitive(PType::U32, NonNullable));
+    /// ```
+    pub fn resolve(&self, mut dtype: DType) -> VortexResult<DType> {
```

Comment:
This feels flipped. A DType should be able to resolve a FieldPath to get back a DType imo?

Reply:
I think of a field path as a regex like thing analogous to str matching, so what ever that implies